Prometheus unavailable alert

moolitayer opened the pull request:
Contains part of https://github.com/ManageIQ/manageiq/issues/16395

The PR will catch the most common cases where Prometheus is unavailable/returns unexpected response. It won't catch errors where the platform does not start the collector - e.g auth failure. When the collection returns to normal, the alert should resolve. This allows seeing alerts in the monitoring screen when Prometheus is unavailable.

moolitayer commented:
@agrare please review - as part of the discussion on event collector unavailable alert. Is this an acceptable partial solution until we have something better?

moolitayer commented on app/models/manageiq/providers/kubernetes/monitoring_manager/event_catcher/stream.rb:
```diff
@@ -81,4 +91,38 @@ def last_position
+  def status_event_common
+    {
+      "annotations" => {
+        "url"       => "www.example.com",
```
This should be a URL that explains how to resolve the error. I think I will take this from settings

moolitayer commented on app/models/manageiq/providers/kubernetes/monitoring_manager/event_catcher/stream.rb:
```diff
@@ -81,4 +91,38 @@ def last_position
+  def collection_compromised_event
+    # TODO: Using options for this end is a hack, find a better solution
+    @ems.parent_manager.options[:last_collection_problem] ||= Time.zone.now
+    @ems.parent_manager.save!
```
The reason we need to persist the start time is that it is used for resolving of alerts:

https://github.com/ManageIQ/manageiq-providers-kubernetes/blob/e3a711af542f8ba68373e3cd5e2e17198c1b3f51/app/models/manageiq/providers/kubernetes/monitoring_manager/event_catcher/runner_mixin.rb#L91

https://github.com/ManageIQ/manageiq/blob/57b5da5bde3227154d4e1173f28cd324ea266b01/app/models/miq_alert.rb#L251

cben replied:
I was just considering something similar to persist last refresh resourceVersion :-)
The migration-less hack I was thinking of is a CustomAttribute.

moolitayer replied:
:+1: That sounds better then using options.

agrare commented on app/models/manageiq/providers/kubernetes/monitoring_manager/event_catcher/stream.rb:
```diff
@@ -81,4 +91,38 @@ def last_position
+      "status" => "resolved"
+    )
+    @ems.parent_manager.options[:last_collection_problem] = nil
+    @ems.parent_manager.save!
```
How often would this save the manager?  On every successful alert?

moolitayer replied:
On every iteration which is currently 20 sec. Good point, I will do it only if changed

moolitayer commented:
@zeari @cben @agrare 

I've pushed a new rebased version and addressed comments.
Honestly I don't like the current solution persisting EmsEvent every queue interval(alerts working).
Also it would have to be almost rewritten after we move to use a streaming API.

I believe raising an EmsEvent might be the solution, I will try to explore that.
The down side is need for policies to be defined. I wonder if we could use the same policy somehow.

moolitayer commented:
I closing this as it is not a complete solution and would have to be removed once we do have one
